### DISCUSSION TITLE
Importing and plotting CSDM data

### DISCUSSION THREAD

**Discussion by pdellis01** [deepanshs/mrsimulator#263] [General]
In all of your examples for host you utilize a web site address ... my data set resides in a directory on my computer ... when I write a path to the file I receive an error message saying there is no such file or directory.  Any suggestions on my error?

**Comment by mgiammar** [comment]
Would you be able to copy and paste the few lines you are using to try and load a .csdm object into this thread? The issue might lie there.

Also, with regards to the `host` and `filename` variables in the online examples, they are just strings which get added together into a single string to load that file from a URL. For example
```
host = "https://nmr.cemhti.cnrs-orleans.fr/Dmfit/Help/csdm/"
filename = "31P Phosphate 6kHz.csdf"
experiment = cp.load(host + filename)
```
is equivalent to calling
```
experiment = cp.load("https://nmr.cemhti.cnrs-orleans.fr/Dmfit/Help/csdm/31P Phosphate 6kHz.csdf")
```

For local files, you won't need a https url, just the path to whatever file. For example,
```
experiment = cp.load("/Users/nmruser/Desktop/some_exp_spectrum.csdf")
```

**Reply by pdellis01** [reply]
Thank you as far as I see the cp.load worked ... then ... when I ran the program (very simple one):

import csdmpy as cp
import matplotlib.pyplot as plt
from lmfit import Minimizer

from mrsimulator import Simulator, SpinSystem, Site
from mrsimulator.method.lib import BlochDecaySpectrum
from mrsimulator.utils import spectral_fitting as sf
from mrsimulator.utils import get_spectral_dimensions
from mrsimulator.spin_system.tensors import SymmetricTensor


# read in the JEOL dataset file


experiment = cp.load ("/Users/pdellis/Desktop/SnO_Better.csdf")
experiment = experiment.real


plt.figure(4.25,3.0)
ax = plt.subplot(projection="csdm")
ax.plt(experiment, colors="black", linewidth=0.5, label ="Experiment")
ax.set_xlim(150,-150)
plt.grid()
plt.show()
 
I then received the following error message:
 /usr/local/bin/python3 /Users/pdellis/Desktop/mrsim_fit/new_SnO.py
pdellis@dotynmr mrsim_fit % /usr/local/bin/python3 /Users/pdellis/Desktop/mrsim_fit/new_SnO.py
Traceback (most recent call last):
  File "/Users/pdellis/Desktop/mrsim_fit/new_SnO.py", line 20, in <module>
    plt.figure(4.25,3.0)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/_api/deprecation.py", line 454, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/pyplot.py", line 840, in figure
    manager = new_figure_manager(
              ^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/pyplot.py", line 384, in new_figure_manager
    return _get_backend_mod().new_figure_manager(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/backend_bases.py", line 3573, in new_figure_manager
    fig = fig_cls(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/_api/deprecation.py", line 454, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/figure.py", line 2536, in __init__
    self.bbox_inches = Bbox.from_bounds(0, 0, *figsize)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: Value after * must be an iterable, not float
pdellis@dotynmr mrsim_fit %

**Reply by mgiammar** [reply]
Looks like just matplotlib issues now. Some of the arguments and functions called don't match what matplotlib is expecting. The following code should work for plotting:

```
plt.figure(figsize=(4.25, 3.0))
ax = plt.subplot(projection="csdm")
ax.plot(experiment, colors="black", linewidth=0.5, label ="Experiment")
ax.set_xlim(150,-150)
plt.grid()
plt.show()
```

Size figure passed as `figsize=(width, height)` to the `plt.figure()` call, cannot just be two numbers.

**Reply by pdellis01** [reply]
we are getting closer ... it did not like the word color:
usr/local/bin/python3 /Users/pdellis/Desktop/mrsim_fit/new_SnO.py
pdellis@dotynmr mrsim_fit % /usr/local/bin/python3 /Users/pdellis/Desktop/mrsim_fit/new_SnO.py
Traceback (most recent call last):
  File "/Users/pdellis/Desktop/mrsim_fit/new_SnO.py", line 20, in <module>
    ax.plot(experiment, colors="black", linewidth=0.5, label ="Experiment")
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/csdmpy/helper_functions.py", line 56, in plot
    return self._call_1D(csdm, "plot", *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/csdmpy/helper_functions.py", line 195, in _call_1D
    r_plt = super().plot(x_, y_, *args, **kwargs_)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/axes/_axes.py", line 1688, in plot
    lines = [*self._get_lines(*args, data=data, **kwargs)]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/axes/_base.py", line 311, in __call__
    yield from self._plot_args(
               ^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/axes/_base.py", line 544, in _plot_args
    return [l[0] for l in result]
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/axes/_base.py", line 544, in <listcomp>
    return [l[0] for l in result]
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/axes/_base.py", line 537, in <genexpr>
    result = (make_artist(x[:, j % ncx], y[:, j % ncy], kw,
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/axes/_base.py", line 351, in _makeline
    seg = mlines.Line2D(x, y, **kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/_api/deprecation.py", line 454, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/lines.py", line 393, in __init__
    self._internal_update(kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/artist.py", line 1223, in _internal_update
    return self._update_props(
           ^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/matplotlib/artist.py", line 1197, in _update_props
    raise AttributeError(
AttributeError: Line2D.set() got an unexpected keyword argument 'colors'
pdellis@dotynmr mrsim_fit %

**Reply by mgiammar** [reply]
Try changing the `colors` argument to just `color`.

**Comment by pdellis01** [comment]
your suggestion worked!  I got a nice blank box ... no spectrum.  The csdf file has an asci header, the data as binary.  Are you expecting asci?

**Reply by mgiammar** [reply]
Remove the line setting the x limits. Matplotlib should automatically set the limits for whatever dataset you're plotting.

**Comment by pdellis01** [comment]
Cool I am at play ground with my grand kids but I will give your suggestions a quick shot.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by pdellis01** [comment]
Still having Fun ...

I still cannot get my simulated spectrum to display ... it is not an issue with plot limits.  The code now runs to completion ... but I am sure it is NOT executing the simulation or the fit.  However, I am getting an error message:

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/astropy/units/quantity.py:673: RuntimeWarning: divide by zero encountered in divide
  result = super().__array_ufunc__(function, method, *arrays, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/astropy/units/quantity.py:673: RuntimeWarning: invalid value encountered in divide

I am attaching the program and one of the plots ...

The Plot:
![Figure_1](https://github.com/deepanshs/mrsimulator/assets/134020204/a5a11377-2452-4a0b-9627-b30ddb80308e)

The code:

from doctest import master
from gettext import ngettext
import locale
import csdmpy as cp
import matplotlib.pyplot as plt
from lmfit import Minimizer
from mrsimulator.spin_system.site import SymmetricTensor

from mrsimulator import Site, SpectralDimension, SpinSystem, Simulator, spin_system
from mrsimulator.method.lib import BlochDecaySpectrum
from mrsimulator import signal_processor as sp
from mrsimulator.utils import spectral_fitting as sf
from mrsimulator.method import SpectralDimension
from mrsimulator.utils import get_spectral_dimensions

# initialize nmrglue converter object 

ax = plt.subplot(projection="csdm")
# read in the JEOL dataset file

experiment = cp.load ("/Users/pdellis/Desktop/mrsim_fit/SnO_Transformed.csdf") 
fig, ax = plt.subplots(1, 2, figsize=(9, 3.5), subplot_kw={"projection": "csdm"})

ax[0].plot(experiment.real)
ax[0].set_title("Full Spectrum")
ax[0].grid()
ax[1].plot(experiment.real,color= "red", label="real")
ax[1].set_title("Zoomed Spectrum")
ax[1].set_xlim(1.4e6, 1.8e6)
ax[1].grid()
plt.tight_layout()
plt.legend()
plt.show()

# standard deviation of noise from the dataset
sigma = 0.6410905

Tin_119 = Site(
    isotope = "119Sn",
    isotropic_chemical_shift = 1.6e6,
    shielding_symmetric = SymmetricTensor(zeta = 1.8e5,eta = 0.0)

    )

Dict = {'origin_offset': 'Hz', 'reference_offset': 'Hz', 'spectral_width': 'Hz', 'reference_offset': 'Hz', 'origin_offset': 'Hz'},
spectral_dims = get_spectral_dimensions(experiment)
bloch = BlochDecaySpectrum(
    channels = ["119Sn"],
    magnetic_flux_density = 11.75,  # in T
    rotor_angle = 0.9553234193716164, # in rad
    rotor_frequency = 6600,  # in Hz  
        spectral_dimensions = [
            SpectralDimension(
            count = 8192,
            spectral_width = 2000000,  # in Hz
        )
        ],
    experiment = experiment,  # add the measurement to the method
    )

# Use only the real part of the spectrum
experiment = experiment.real
sys = [SpinSystem(sites=[Tin_119]), ]

sim = Simulator(spin_systems = sys, methods=[bloch])


sim.config.decompose_spectrum = "spin_system"

sim.run()

relative_intensity_factor = experiment.max() / sim.methods[0].simulation.max()

# Post Simulation Processing
processor = sp.SignalProcessor(
    operations=[
        sp.IFFT(),
        sp.apodization.Gaussian(FWHM="50 Hz"),
        sp.FFT(),        
        sp.Scale(factor=200)
    ]
)
dataset = experiment
processed_dataset = processor.apply_operations(dataset = sim.methods[0].simulation).real
# Plot the spectrum
plt.figure(figsize=(4.25, 3.0))
ax = plt.subplot(projection="csdm")
ax.plot(experiment, linewidth=1, label="Experiment")
ax = plt.subplot(projection="csdm")
ax.plot(processed_dataset, linewidth=1, label="Simulation")
ax.set_xlim(1.4e6, 1.8e6)
plt.grid()
plt.legend()
plt.tight_layout()
plt.show()


params = sf.make_LMFIT_params(sim, processor, include={"rotor_frequency"})

fit_parameters = sf.make_LMFIT_params(sim, processor)
print(fit_parameters.pretty_print(columns=["value", "min", "max", "vary", "expr"]))

# standard deviation of noise from the dataset
from lmfit import Minimizer
minner = Minimizer(sf.LMFIT_min_function, fit_parameters, fcn_args=(sim, processor, sigma))
result = minner.minimize()
result

best_fit = sf.bestfit(sim, processor)[0].real
residuals = sf.residuals(sim, processor)[0].real

# Plot the spectrum
plt.figure(figsize=(4.25, 3.0))
ax = plt.subplot(projection="csdm")
ax.plot(experiment, linewidth=1, label="Experiment")
ax.plot(best_fit, linewidth=1, label="Best Fit")
ax.plot(residuals, linewidth=1, label="Residuals")
ax.set_xlim(1.4e6, 1.8e6)
plt.grid()
plt.legend()
plt.tight_layout()
plt.show()

**Reply by pdellis01** [reply]
WE are making progress --- Almost ... No error messages here is the output and  I have attached the program and here are some plots:

pdellis@dotynmr ~ % /usr/local/bin/python3 /Users/pdellis/Desktop/mrsim_fit/SnO_Fitting_RMN.py
2023-06-25 14:42:50.758 Python[29615:723980] WARNING: Secure coding is not enabled for restorable state! Enable secure coding by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState: and returning YES.
Name                                      Value      Min      Max     Vary     Expr
SP_0_operation_1_Gaussian_FWHM                5     -inf      inf     True     None
SP_0_operation_3_Scale_factor              5000     -inf      inf     True     None
sys_0_abundance                             100        0      100    False      100
sys_0_site_0_isotropic_chemical_shift      -650     -inf      inf     True     None
sys_0_site_0_shielding_symmetric_eta          0        0        1     True     None
sys_0_site_0_shielding_symmetric_zeta       600     -inf      inf     True     None
None
pdellis@dotynmr ~ % 

For some reason I am having trouble pasting the spectra ... but you will see the issue when you run the program ...


[SnO_Fitting_RMN.zip](https://github.com/deepanshs/mrsimulator/files/11860555/SnO_Fitting_RMN.zip)

**Reply by mgiammar** [reply]
Hi Paul,

I don't have the file `SnO_Phased-1041.csdf`, so I am unable to run the program. Are you able to attach it?

**Reply by pdellis01** [reply]
Here is the file ...

SnO_Phased-1041.csdf

**Reply by mgiammar** [reply]
Can you compress the file to zip, then drag and drop into the comment box? Otherwise it will not upload to GitHub.

**Reply by pdellis01** [reply]
Here you go ...
[SnO_Phased-1041.zip](https://github.com/deepanshs/mrsimulator/files/11870557/SnO_Phased-1041.zip)

**Comment by mgiammar** [comment]
Find attached a Jupyter Notebook (in a zip file) which runs a fit on the spectrum you provided. You can run this in [Google Colab](https://colab.research.google.com/); the script will install the requisite libraries. Note that the fit is not great mostly because I have not had time to play around with the starting parameters.

You are not seeing the simulated spectrum in the script above since the values for `isotropic_chemical_shift` and `shielding_symmetric.zeta` are still (presumably) in Hz. Mrsimulator *does not* accept these values in Hz; they must a ppm value. The simulated spectrum is therefore on the order of 10^6 ppm, far outside the spectral window of the experimental dataset, which is plotted first.

The sidebands in the spectrum also seem to have no width, that is, they are single points. Could you provide the raw JOEL time-domain dataset? Maybe there is an export discrepancy on their end or some pre-processing issue.

[SnO_Fit.ipynb.zip](https://github.com/deepanshs/mrsimulator/files/11697001/SnO_Fit.ipynb.zip)

**Comment by pdellis01** [comment]
Oh Boy … thank you very much.  I will send you the file in the morning … it is on another computer.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by pdellis01** [comment]
Hi Mathew …

I have attached the time domain file.  I have my day tied up for a while I will send the raw data after lunch

Paul

From: Matthew Giammar ***@***.***>
Date: Thursday, June 8, 2023 at 7:22 PM
To: deepanshs/mrsimulator ***@***.***>
Cc: Paul Ellis ***@***.***>, Author ***@***.***>
Subject: Re: [deepanshs/mrsimulator] Importing and plotting CSDM data (Discussion #263)

Find attached a Jupyter Notebook (in a zip file) which runs a fit on the spectrum you provided. You can run this in Google Colab<https://colab.research.google.com/>; the script will install the requisite libraries. Note that the fit is not great mostly because I have not had time to play around with the starting parameters.

You are not seeing the simulated spectrum in the script above since the values for isotropic_chemical_shift and shielding_symmetric.zeta are still (presumably) in Hz. Mrsimulator does not accept these values in Hz; they must a ppm value. The simulated spectrum is therefore on the order of 10^6 ppm, far outside the spectral window of the experimental dataset, which is plotted first.

The sidebands in the spectrum also seem to have no width, that is, they are single points. Could you provide the raw JOEL time-domain dataset? Maybe there is an export discrepancy on their end or some pre-processing issue.

SnO_Fit.ipynb.zip<https://github.com/deepanshs/mrsimulator/files/11697001/SnO_Fit.ipynb.zip>

**Reply by pdellis01** [reply]
Hi Mathew ...

Some progress ... Mike and I are on the same page (art least the same chapter) ... given that I am concerned about two of the error message I now have ... (see below)

pdellis@dotynmr ~ % /usr/local/bin/python3 /Users/pdellis/Desktop/mrsim_fit/SnO_Fitting.py
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/core/_methods.py:269: RuntimeWarning: Degrees of freedom <= 0 for slice
  ret = _var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/core/_methods.py:226: RuntimeWarning: invalid value encountered in divide
  arrmean = um.true_divide(arrmean, div, out=arrmean,
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/core/_methods.py:261: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
Name                                      Value      Min      Max     Vary     Expr
SP_0_operation_1_Gaussian_FWHM               10     -inf      inf     True     None
SP_0_operation_3_Scale_factor              4000     -inf      inf     True     None
sys_0_abundance                             100        0      100    False      100
sys_0_site_0_isotropic_chemical_shift      -650     -inf      inf     True     None
sys_0_site_0_shielding_symmetric_eta          0        0        1     True     None
sys_0_site_0_shielding_symmetric_zeta       700     -inf      inf     True     None
None
Traceback (most recent call last):
  File "/Users/pdellis/Desktop/mrsim_fit/SnO_Fitting.py", line 139, in <module>
    result = minner.minimize()
             ^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/lmfit/minimizer.py", line 2379, in minimize
    return function(**kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/lmfit/minimizer.py", line 1703, in leastsq
    lsout = scipy_leastsq(self.__residual, variables, **lskws)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/scipy/optimize/_minpack_py.py", line 413, in leastsq
    shape, dtype = _check_func('leastsq', 'func', func, x0, args, n)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/scipy/optimize/_minpack_py.py", line 26, in _check_func
    res = atleast_1d(thefunc(*((x0[:numinputs],) + args)))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/lmfit/minimizer.py", line 604, in __residual
    return _nan_policy(np.asfarray(out).ravel(),
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/lmfit/minimizer.py", line 2463, in _nan_policy
    raise ValueError(msg)
ValueError: NaN values detected in your input data or the output of your objective/model function - fitting algorithms cannot handle this! Please read https://lmfit.github.io/lmfit-py/faq.html#i-get-errors-from-nan-in-my-fit-what-can-i-do for more information.
pdellis@dotynmr ~ % 

Clearly, I cannot see where the where the specific error is occurring ... but any advice would be welocome.  I have copied the offending code below ...

Paul

import csdmpy as cp
import mrsimulator
from mrsimulator import SpectralDimension, signal_processor as sp
from mrsimulator import Site, SpinSystem, Simulator
from mrsimulator.method.lib import BlochDecaySpectrum
from mrsimulator.utils import get_spectral_dimensions
from mrsimulator.spin_system.tensors import SymmetricTensor
from mrsimulator.method import Isotope

import matplotlib.pyplot as plt

import os
os.getcwd()

import math
import cmath

import numpy as np

from lmfit import Minimizer

# read in the JEOL dataset file
testdata = cp.load ("/Users/pdellis/Desktop/mrsim_fit/SnO_Phased-1041.csdf")
testdata.description
x = testdata.dimensions
x


#As saved
plt.figure(figsize=(4.25, 3.0))
ax = plt.subplot(projection="csdm")
ax.plot(testdata.real,color= "black", label="real")
ax.plot(testdata.imag,color= "red", label="Imaginary")
ax.grid()
plt.tight_layout()
plt.legend()
plt.show()

# Reverse the data without reversing the axis

testdata_swap = testdata.real - cmath.sqrt(-1) * testdata.imag

processor = sp.SignalProcessor(
    operations=[sp.apodization.Exponential(FWHM="10 Hz"), sp.FFT()]
)
processed_dataset = processor.apply_operations(dataset=testdata_swap)
processed_dataset.dimensions[0].to("ppm", "nmr_frequency_ratio")

x = processed_dataset.dimensions
x

#As Calculated / Collected
plt.figure(figsize=(4.25, 3.0))
ax = plt.subplot(projection="csdm")
ax.plot(processed_dataset.real, color="black", linewidth=1)
#   ax.invert_xaxis()
ax.set_xlim(0, -1.7E3)
plt.tight_layout()
plt.show()

# Use only the real part of the spectrum
exp_spectrum = processed_dataset.real

# Use region below -20 ppm to calculate the noise standard deviation
loc = np.where(exp_spectrum.dimensions[0].coordinates < -2000)
sigma = exp_spectrum[loc].std()

test = np.abs(processed_dataset)

#As Calculated / Collected
plt.figure(figsize=(4.25, 3.0))
ax = plt.subplot(projection="csdm")
ax.plot(test, color="black", linewidth=1)
#ax.invert_xaxis()
ax.set_xlim(0, -1700)
plt.tight_layout()
plt.show()

site = Site(
    isotope = "119Sn",
    isotropic_chemical_shift = -650.0,
    shielding_symmetric = {"zeta" : 700,"eta" : 0.0},
)

# Get the dimensions from the experimental dataset
spectral_dims = get_spectral_dimensions(test)
v_ref = test.dimensions[0].origin_offset - exp_spectrum.dimensions[0].coordinates_offset
B0 = v_ref.to("MHz").value / np.abs(Isotope(symbol="119Sn").gyromagnetic_ratio)

MAS =  BlochDecaySpectrum(
    channels=["119Sn"],
    magnetic_flux_density = 11.75,  # in T
    rotor_frequency = 8067,  # in Hz
    spectral_dimensions = spectral_dims,

    experiment = test,  # add the measurement to the method.Method
)

sys = SpinSystem(sites = [site])

sim = Simulator(spin_systems = [sys], methods = [MAS])
sim.config.number_of_sidebands = 60
sim.run()

# Post Simulation Processing
# --------------------------
relative_intensity_factor = exp_spectrum.max() / sim.methods[0].simulation.max()
processor = sp.SignalProcessor(
    operations=[
        sp.IFFT(),
        sp.apodization.Gaussian(FWHM="10 Hz"),
        sp.FFT(),
        sp.Scale(factor=4000),
    ]
)
processed_dataset = processor.apply_operations(dataset=sim.methods[0].simulation).real
# Plot of the guess spectrum
# --------------------------
plt.figure(figsize=(6.0, 3.0))
ax = plt.subplot(projection="csdm")
ax.plot(test, label="Experiment")
ax.plot(processed_dataset, label="Guess Spectrum")
ax.set_xlim(0, -1700)
plt.legend()
plt.grid()
plt.tight_layout()
plt.show()


from mrsimulator.utils import spectral_fitting as sf
fit_parameters = sf.make_LMFIT_params(sim, processor)
print(fit_parameters.pretty_print(columns=["value", "min", "max", "vary", "expr"]))
fit_parameters["sys_0_site_0_shielding_symmetric_eta"].value = 0
fit_parameters["sys_0_site_0_shielding_symmetric_zeta"].vary = 0
fit_parameters["sys_0_site_0_isotropic_chemical_shift"].vary = 0

from lmfit import Minimizer
minner = Minimizer(sf.LMFIT_min_function, fit_parameters, fcn_args=(sim, processor, sigma))
result = minner.minimize()
result

best_fit = sf.bestfit(sim, processor)[0].real
residuals = sf.residuals(sim, processor)[0].real

# Plot the spectrum
plt.figure(figsize=(6, 3.0))
ax = plt.subplot(projection="csdm")
ax.plot(exp_spectrum, label="Experiment")
ax.plot(best_fit, alpha=0.75, label="Best Fit")
ax.plot(residuals, alpha=0.75, label="Residuals")
ax.set_xlim(-15, 15)
plt.legend()
plt.grid()
plt.tight_layout()
plt.show()
fit_parameters["sys_0_site_0_shielding_symmetric_eta"].value = 0
fit_parameters["sys_0_site_0_shielding_symmetric_eta"].vary = True
minner = Minimizer(sf.LMFIT_min_function, fit_parameters, fcn_args=(sim, processor, sigma))
result = minner.minimize()
best_fit = sf.bestfit(sim, processor)[0].real
residuals = sf.residuals(sim, processor)[0].real

# Plot the spectrum
plt.figure(figsize=(6, 3.0))
ax = plt.subplot(projection="csdm")
ax.plot(exp_spectrum, label="Experiment")
ax.plot(best_fit, alpha=0.75, label="Best Fit")
ax.plot(residuals, alpha=0.75, label="Residuals")
ax.set_xlim(-15, 15)
plt.legend()
plt.grid()
plt.tight_layout()
plt.show()

**Comment by pdellis01** [comment]
Double darn  … I will send right away.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by mgiammar** [comment]
@pdellis01 Attached is a zip file with a notebook and a processed 1D static spectrum of your SnO CPMG dataset. The notebook sets up a static Method object to simulate the experimental spectrum and subsequently fits the dataset; the residuals don't look too bad.

I think your issues are coming from trying to fit a spikelet pattern from a 1D echo train to a spinning sideband simulation. I don't know if this is possible, and mrsimulator is certainly not designed to handle such a case.

Hopefully this notebook helps!

[SnO_fit_plus_dataset.zip](https://github.com/deepanshs/mrsimulator/files/11884553/SnO_fit_plus_dataset.zip)

**Comment by pdellis01** [comment]
Mathew …

I can’t thank you enough …

Warmest regards …

Paul

From: Matthew Giammar ***@***.***>
Date: Tuesday, June 27, 2023 at 11:38 AM
To: deepanshs/mrsimulator ***@***.***>
Cc: Paul Ellis ***@***.***>, Mention ***@***.***>
Subject: Re: [deepanshs/mrsimulator] Importing and plotting CSDM data (Discussion #263)

@pdellis01<https://github.com/pdellis01> Attached is a zip file with a notebook and a processed 1D static spectrum of your SnO CPMG dataset. The notebook sets up a static Method object to simulate the experimental spectrum and subsequently fits the dataset; the residuals don't look too bad.

I think your issues are coming from trying to fit a spikelet pattern from a 1D echo train to a spinning sideband simulation. I don't know if this is possible, and mrsimulator is certainly not designed to handle such a case.

Hopefully this notebook helps!

SnO_fit_plus_dataset.zip<https://github.com/deepanshs/mrsimulator/files/11884553/SnO_fit_plus_dataset.zip>
